File: snippets/algorithms/graph-traversal/graph_traversal.php
<?php
/**
 * Graph Traversal Algorithms in PHP
 * 
 * This script demonstrates BFS and DFS traversal algorithms on a graph.
 * 
 * Run: php graph_traversal.php
 */

/**
 * A graph class using adjacency list representation
 */
class Graph {
    // Adjacency list representation
    private $adjacencyList = [];
    
    /**
     * Adds a vertex to the graph
     * 
     * @param string $vertex The vertex to add
     */
    public function addVertex($vertex) {
        if (!isset($this->adjacencyList[$vertex])) {
            $this->adjacencyList[$vertex] = [];
        }
    }
    
    /**
     * Adds an edge between two vertices
     * 
     * @param string $v1 First vertex
     * @param string $v2 Second vertex
     */
    public function addEdge($v1, $v2) {
        // Ensure both vertices exist
        $this->addVertex($v1);
        $this->addVertex($v2);
        
        // Add the edge (undirected graph)
        $this->adjacencyList[$v1][] = $v2;
        $this->adjacencyList[$v2][] = $v1;
    }
    
    /**
     * Helper method to get sorted neighbors for consistent output
     * 
     * @param string $vertex The vertex to get neighbors for
     * @return array Sorted array of neighbors
     */
    private function getSortedNeighbors($vertex) {
        $neighbors = $this->adjacencyList[$vertex];
        sort($neighbors);
        return $neighbors;
    }
    
    /**
     * Performs a breadth-first search traversal starting from the given vertex
     * 
     * @param string $start Starting vertex
     * @return array Array of vertices in the order they were visited
     */
    public function bfs($start) {
        if (!isset($this->adjacencyList[$start])) {
            return [];
        }
        
        $visited = [$start => true];
        $queue = [$start];
        $result = [];
        
        echo "Starting BFS traversal from vertex $start\n";
        
        while (count($queue) > 0) {
            // Dequeue the first vertex
            $vertex = array_shift($queue);
            $result[] = $vertex;
            
            echo "Visiting: $vertex\n";
            echo "Queue: [" . implode(", ", $queue) . "]\n";
            echo "Visited so far: [" . implode(", ", $result) . "]\n";
            echo "------------------------------\n";
            
            // Pause for demonstration
            usleep(500000); // 500ms
            
            // Get sorted neighbors for consistent order
            $neighbors = $this->getSortedNeighbors($vertex);
            
            // Enqueue all unvisited neighbors
            foreach ($neighbors as $neighbor) {
                if (!isset($visited[$neighbor])) {
                    $visited[$neighbor] = true;
                    $queue[] = $neighbor;
                }
            }
        }
        
        return $result;
    }
    
    /**
     * Performs a recursive depth-first search traversal starting from the given vertex
     * 
     * @param string $start Starting vertex
     * @return array Array of vertices in the order they were visited
     */
    public function dfsRecursive($start) {
        if (!isset($this->adjacencyList[$start])) {
            return [];
        }
        
        $visited = [];
        $result = [];
        
        echo "Starting recursive DFS traversal from vertex $start\n";
        
        $this->dfsHelper($start, $visited, $result);
        
        return $result;
    }
    
    /**
     * Helper method for recursive DFS
     * 
     * @param string $vertex Current vertex
     * @param array &$visited Reference to array of visited vertices
     * @param array &$result Reference to array to store the traversal result
     */
    private function dfsHelper($vertex, &$visited, &$result) {
        // Mark as visited and add to result
        $visited[$vertex] = true;
        $result[] = $vertex;
        
        echo "Visiting: $vertex\n";
        echo "Visited so far: [" . implode(", ", $result) . "]\n";
        echo "------------------------------\n";
        
        // Pause for demonstration
        usleep(500000); // 500ms
        
        // Get sorted neighbors for consistent order
        $neighbors = $this->getSortedNeighbors($vertex);
        
        // Recursively visit all unvisited neighbors
        foreach ($neighbors as $neighbor) {
            if (!isset($visited[$neighbor])) {
                $this->dfsHelper($neighbor, $visited, $result);
            }
        }
    }
    
    /**
     * Performs an iterative depth-first search traversal starting from the given vertex
     * 
     * @param string $start Starting vertex
     * @return array Array of vertices in the order they were visited
     */
    public function dfsIterative($start) {
        if (!isset($this->adjacencyList[$start])) {
            return [];
        }
        
        $visited = [];
        $stack = [$start];
        $result = [];
        
        echo "Starting iterative DFS traversal from vertex $start\n";
        
        while (count($stack) > 0) {
            // Pop the top vertex
            $vertex = array_pop($stack);
            
            // If not visited, process it
            if (!isset($visited[$vertex])) {
                $visited[$vertex] = true;
                $result[] = $vertex;
                
                echo "Visiting: $vertex\n";
                echo "Stack: [" . implode(", ", $stack) . "]\n";
                echo "Visited so far: [" . implode(", ", $result) . "]\n";
                echo "------------------------------\n";
                
                // Pause for demonstration
                usleep(500000); // 500ms
                
                // Get sorted neighbors in reverse order for stack
                $neighbors = $this->getSortedNeighbors($vertex);
                $neighbors = array_reverse($neighbors);
                
                // Push all unvisited neighbors onto the stack
                foreach ($neighbors as $neighbor) {
                    if (!isset($visited[$neighbor])) {
                        $stack[] = $neighbor;
                    }
                }
            }
        }
        
        return $result;
    }
    
    /**
     * Prints a visualization of the graph structure
     */
    public function visualizeGraph() {
        echo "\nGraph Structure:\n";
        echo "------------------------------\n";
        
        // Sort vertices for consistent output
        $vertices = array_keys($this->adjacencyList);
        sort($vertices);
        
        foreach ($vertices as $vertex) {
            $neighbors = $this->getSortedNeighbors($vertex);
            echo "$vertex -> [" . implode(", ", $neighbors) . "]\n";
        }
        
        echo "------------------------------\n";
    }
}

/**
 * Creates a sample graph for demonstration
 * 
 * @return Graph A sample graph
 */
function createSampleGraph() {
    $g = new Graph();
    
    // Add edges to build this graph:
    //     A
    //    / \
    //   B   C
    //  / \   \
    // D   E---F
    
    $edges = [
        ["A", "B"], ["A", "C"],
        ["B", "D"], ["B", "E"],
        ["C", "F"], ["E", "F"]
    ];
    
    foreach ($edges as $edge) {
        $g->addEdge($edge[0], $edge[1]);
    }
    
    return $g;
}

// Main execution
$g = createSampleGraph();
$g->visualizeGraph();

// Demonstrate BFS
echo "\n=== BFS Traversal ===\n";
$bfsResult = $g->bfs("A");
echo "BFS Result: [" . implode(", ", $bfsResult) . "]\n";

// Demonstrate recursive DFS
echo "\n=== DFS Traversal (Recursive) ===\n";
$dfsRecResult = $g->dfsRecursive("A");
echo "DFS Recursive Result: [" . implode(", ", $dfsRecResult) . "]\n";

// Demonstrate iterative DFS
echo "\n=== DFS Traversal (Iterative) ===\n";
$dfsIterResult = $g->dfsIterative("A");
echo "DFS Iterative Result: [" . implode(", ", $dfsIterResult) . "]\n";
?> 

File: snippets/algorithms/graph-traversal/GraphTraversal.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;

/**
 * Graph Traversal Algorithms in C#
 * 
 * This program demonstrates BFS and DFS traversal algorithms on a graph.
 * 
 * Compile: dotnet build
 * Run: dotnet run
 */
namespace GraphTraversalAlgorithms
{
    /// <summary>
    /// A graph class using adjacency list representation
    /// </summary>
    public class Graph
    {
        // Adjacency list representation
        private Dictionary<string, List<string>> adjacencyList;

        /// <summary>
        /// Initializes a new empty graph
        /// </summary>
        public Graph()
        {
            adjacencyList = new Dictionary<string, List<string>>();
        }

        /// <summary>
        /// Adds a vertex to the graph
        /// </summary>
        /// <param name="vertex">The vertex to add</param>
        public void AddVertex(string vertex)
        {
            if (!adjacencyList.ContainsKey(vertex))
            {
                adjacencyList[vertex] = new List<string>();
            }
        }

        /// <summary>
        /// Adds an edge between two vertices
        /// </summary>
        /// <param name="v1">First vertex</param>
        /// <param name="v2">Second vertex</param>
        public void AddEdge(string v1, string v2)
        {
            // Ensure both vertices exist
            AddVertex(v1);
            AddVertex(v2);

            // Add the edge (undirected graph)
            adjacencyList[v1].Add(v2);
            adjacencyList[v2].Add(v1);
        }

        /// <summary>
        /// Helper method to get sorted neighbors for consistent output
        /// </summary>
        /// <param name="vertex">The vertex to get neighbors for</param>
        /// <returns>Sorted list of neighbors</returns>
        private List<string> GetSortedNeighbors(string vertex)
        {
            var neighbors = new List<string>(adjacencyList[vertex]);
            neighbors.Sort();
            return neighbors;
        }

        /// <summary>
        /// Performs a breadth-first search traversal starting from the given vertex
        /// </summary>
        /// <param name="start">Starting vertex</param>
        /// <returns>List of vertices in the order they were visited</returns>
        public List<string> BFS(string start)
        {
            if (!adjacencyList.ContainsKey(start))
            {
                return new List<string>();
            }

            HashSet<string> visited = new HashSet<string>();
            Queue<string> queue = new Queue<string>();
            List<string> result = new List<string>();

            // Initialize with starting vertex
            visited.Add(start);
            queue.Enqueue(start);

            Console.WriteLine($"Starting BFS traversal from vertex {start}");

            while (queue.Count > 0)
            {
                // Dequeue the first vertex
                string vertex = queue.Dequeue();
                result.Add(vertex);

                Console.WriteLine($"Visiting: {vertex}");
                Console.WriteLine($"Queue: [{string.Join(", ", queue)}]");
                Console.WriteLine($"Visited so far: [{string.Join(", ", result)}]");
                Console.WriteLine("------------------------------");

                // Pause for demonstration
                Thread.Sleep(500);

                // Get sorted neighbors for consistent order
                List<string> neighbors = GetSortedNeighbors(vertex);

                // Enqueue all unvisited neighbors
                foreach (string neighbor in neighbors)
                {
                    if (!visited.Contains(neighbor))
                    {
                        visited.Add(neighbor);
                        queue.Enqueue(neighbor);
                    }
                }
            }

            return result;
        }

        /// <summary>
        /// Performs a recursive depth-first search traversal starting from the given vertex
        /// </summary>
        /// <param name="start">Starting vertex</param>
        /// <returns>List of vertices in the order they were visited</returns>
        public List<string> DFSRecursive(string start)
        {
            if (!adjacencyList.ContainsKey(start))
            {
                return new List<string>();
            }

            HashSet<string> visited = new HashSet<string>();
            List<string> result = new List<string>();

            Console.WriteLine($"Starting recursive DFS traversal from vertex {start}");

            DFSHelper(start, visited, result);

            return result;
        }

        /// <summary>
        /// Helper method for recursive DFS
        /// </summary>
        /// <param name="vertex">Current vertex</param>
        /// <param name="visited">Set of visited vertices</param>
        /// <param name="result">List to store the traversal result</param>
        private void DFSHelper(string vertex, HashSet<string> visited, List<string> result)
        {
            // Mark as visited and add to result
            visited.Add(vertex);
            result.Add(vertex);

            Console.WriteLine($"Visiting: {vertex}");
            Console.WriteLine($"Visited so far: [{string.Join(", ", result)}]");
            Console.WriteLine("------------------------------");

            // Pause for demonstration
            Thread.Sleep(500);

            // Get sorted neighbors for consistent order
            List<string> neighbors = GetSortedNeighbors(vertex);

            // Recursively visit all unvisited neighbors
            foreach (string neighbor in neighbors)
            {
                if (!visited.Contains(neighbor))
                {
                    DFSHelper(neighbor, visited, result);
                }
            }
        }

        /// <summary>
        /// Performs an iterative depth-first search traversal starting from the given vertex
        /// </summary>
        /// <param name="start">Starting vertex</param>
        /// <returns>List of vertices in the order they were visited</returns>
        public List<string> DFSIterative(string start)
        {
            if (!adjacencyList.ContainsKey(start))
            {
                return new List<string>();
            }

            HashSet<string> visited = new HashSet<string>();
            Stack<string> stack = new Stack<string>();
            List<string> result = new List<string>();

            // Initialize with starting vertex
            stack.Push(start);

            Console.WriteLine($"Starting iterative DFS traversal from vertex {start}");

            while (stack.Count > 0)
            {
                // Pop the top vertex
                string vertex = stack.Pop();

                // If not visited, process it
                if (!visited.Contains(vertex))
                {
                    visited.Add(vertex);
                    result.Add(vertex);

                    Console.WriteLine($"Visiting: {vertex}");
                    Console.WriteLine($"Stack: [{string.Join(", ", stack)}]");
                    Console.WriteLine($"Visited so far: [{string.Join(", ", result)}]");
                    Console.WriteLine("------------------------------");

                    // Pause for demonstration
                    Thread.Sleep(500);

                    // Get sorted neighbors in reverse order for stack
                    List<string> neighbors = GetSortedNeighbors(vertex);
                    neighbors.Reverse();

                    // Push all unvisited neighbors onto the stack
                    foreach (string neighbor in neighbors)
                    {
                        if (!visited.Contains(neighbor))
                        {
                            stack.Push(neighbor);
                        }
                    }
                }
            }

            return result;
        }

        /// <summary>
        /// Prints a visualization of the graph structure
        /// </summary>
        public void VisualizeGraph()
        {
            Console.WriteLine("\nGraph Structure:");
            Console.WriteLine("------------------------------");

            // Sort vertices for consistent output
            var vertices = adjacencyList.Keys.ToList();
            vertices.Sort();

            foreach (string vertex in vertices)
            {
                List<string> neighbors = GetSortedNeighbors(vertex);
                Console.WriteLine($"{vertex} -> [{string.Join(", ", neighbors)}]");
            }

            Console.WriteLine("------------------------------");
        }
    }

    public class Program
    {
        /// <summary>
        /// Creates a sample graph for demonstration
        /// </summary>
        /// <returns>A sample graph</returns>
        public static Graph CreateSampleGraph()
        {
            Graph g = new Graph();

            // Add edges to build this graph:
            //     A
            //    / \
            //   B   C
            //  / \   \
            // D   E---F

            string[,] edges = {
                {"A", "B"}, {"A", "C"},
                {"B", "D"}, {"B", "E"},
                {"C", "F"}, {"E", "F"}
            };

            for (int i = 0; i < edges.GetLength(0); i++)
            {
                g.AddEdge(edges[i, 0], edges[i, 1]);
            }

            return g;
        }

        public static void Main(string[] args)
        {
            // Create a sample graph
            Graph g = CreateSampleGraph();
            g.VisualizeGraph();

            // Demonstrate BFS
            Console.WriteLine("\n=== BFS Traversal ===");
            List<string> bfsResult = g.BFS("A");
            Console.WriteLine($"BFS Result: [{string.Join(", ", bfsResult)}]");

            // Demonstrate recursive DFS
            Console.WriteLine("\n=== DFS Traversal (Recursive) ===");
            List<string> dfsRecResult = g.DFSRecursive("A");
            Console.WriteLine($"DFS Recursive Result: [{string.Join(", ", dfsRecResult)}]");

            // Demonstrate iterative DFS
            Console.WriteLine("\n=== DFS Traversal (Iterative) ===");
            List<string> dfsIterResult = g.DFSIterative("A");
            Console.WriteLine($"DFS Iterative Result: [{string.Join(", ", dfsIterResult)}]");
        }
    }
} 